kernel/src/transaction.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::actions::visitors::TransactionVisitor as TxnActionVisitor;
+use crate::actions::{get_log_schema, TRANSACTION_NAME};
+use crate::engine_data::TypedGetData;
+use crate::snapshot::Snapshot;
+use crate::EngineInterface;
+use crate::{actions::Transaction, engine_data::GetData, DataVisitor, DeltaResult};
+
+pub type TransactionMap = HashMap<String, Transaction>;
+pub struct TransactionScanner<'a> {
+    snapshot: Arc<Snapshot>,
+    engine: &'a dyn EngineInterface,
+}
+
+impl<'a> TransactionScanner<'a> {
+    pub fn new(engine: &'a dyn EngineInterface, snapshot: Arc<Snapshot>) -> Self {
+        TransactionScanner { snapshot, engine }
+    }
+
+    fn _application_transactions(
```

Comment:
nit: we don't currently have a convention of naming private stuff with an `_`, so I'd just call this `application_transactions` for now

Reply:
`application_transactions` already exists for this struct so I just renamed it to `scan_application_transactions`. 
When it comes to the current implementation I wanted to support obtaining the latest transaction of an application and all latest transactions. 

During implementation I had two designs in mind, either build different visitors for both use cases or just one visitor. Obtaining the latest transaction for an app id could be improved since a hash map does not need to be maintained in that instance but it would likely result in very small gains in speed / space.